keps/2349-multikueue-external-custom-job-support/README.md

```diff
@@ -0,0 +1,186 @@
+title can help communicate what the KEP is and should be considered as part of
+any review.
+-->
+
+<!--
+A table of contents is helpful for quickly jumping to sections of a KEP and for
+highlighting any additional information provided beyond the standard KEP
+template.
+
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+  - [Constraints](#constraints)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This KEP introduces the ability to create Multikueue Admission Check controllers with custom set of adapters.
+
+## Motivation
+
+Many Kueue adaptors have company specific CustomResources.
+Kueue should give a possibility to manage such CustomResources across multiple clusters.
+
+### Goals
+
+Give a possibility for external controllers to manage private CustomResources in a MultiKueue environment.
+
+### Non-Goals
+
+## Proposal
+
+Make multiple instances of Multikueue admission check controller able to run simultaneously with different set of Multkiueue adapters in one cluster at the same time.
+Provide a custom admission check controller name and adapters for Multikueue.
+Specified controller is used by workload, cluster and admission check MultiKueue reconcilers.
+
+### User Stories
+
+#### Story 1
+
+I would like to support the extension mechanism so that we can implement the MultiKueue controller for the custom / in-house Jobs.
+
+### Constraints
+
+Internally the MultiKueue uses a set of controller-runtime client, one for each MultiKueueCluster, the clients are used both by the MultiKueue internals and passed to the job adapters for the custom jobs synchronization.
+
+The clients cannot be share with an external controller and assuming the external controller can get the same connectivity to the worker clusters might be wrong, especially when we are taking about FS mounted kubeconfigs.
+
+Designing additional APIs to communicate between the internal and external controller will bring extra complexity and latency at runtime. 
+
+### Risks and Mitigations
+
+## Design Details
+
+Provide a way to integrate external controllers to manage Jobs in the Kueue.
+MulikueueCluster API needs to be modified to ensure that one multikueue cluster is only managed by one admission check controller.
+
+1. Add ControllerName to the `MultiKueueClusterSpec`
```

Comment:
This is a non-starting point for this proposal.

Asking users to create one MultiKueueCluster object for each integration is too much burden.

Sure, they just need one for all of their CRDs, but what if they need to support a native jobs + in-house jobs + third-party jobs that offer MK support. Now they have at least 3 configurations for the same thing.

How does a MultiKueueConfig that supports all of these jobs look like?

IIUC, your intention with a separate object is to have conditions separate for each external admission check. Then let's maybe add an object just for that, that only one controller processes each time, separate from MultiKueueCluster.

Reply:
One alternative to this, described later, is tho have a custom condition type for the MultiKueueCluster that reflects the connection stale for each ControllerName using it.

> Asking users to create one MultiKueueCluster object for each integration is too much burden.

It should not be for each integration, is for each configured controller , the controller can support any number of adapters. 

> How does a MultiKueueConfig that supports all of these jobs look like?

Nothing changes for MultiKueueConfig, they are just used by an AdmissionCheck (which already has a ControllerName)

Reply:
> One alternative to this, described later, is to have a custom condition type for the MultiKueueCluster that reflects the connection stale for each ControllerName using it.

This sounds much better.